recording/src/nextcloud/talk/recording/Participant.py

```diff
@@ -202,6 +209,50 @@ def __del__(self):
             # created in "/tmp".
             self.driver.quit()
 
+    def startChrome(self, width, height, env):
+        """
+        Starts a Chrome instance. Will use Chromium is Google Chrome is not installed.
+        :param width: the width of the browser window.
+        :param height: the height of the browser window.
+        :param env: the environment variables, including the display to start
+                    the browser in.
+        """
+
+        options = ChromeOptions()
+
+        # "webSocketUrl" is needed for BiDi; this should be set already by
+        # default, but just in case.
+        options.set_capability('webSocketUrl', True)
+
+        options.add_argument('--use-fake-ui-for-media-stream')
+
+        # Allow to play media without user interaction.
+        options.add_argument('--autoplay-policy=no-user-gesture-required')
+
+        options.add_argument('--kiosk')
+        options.add_argument(f'--window-size={width},{height}')
+        options.add_argument('--disable-infobars')
+        options.add_argument('--no-sandbox')
```

Comment:
While I trust Talk :-) I would prefer not to disable the sandbox. Why is that needed?

Reply:
I was unable to run Chromium in my test environment without it. ARM64 + Rocky Linux 9 + Docker.

Refs:
- https://github.com/docker/for-linux/issues/496
- https://github.com/jessfraz/dockerfiles/issues/65

Reply:
I've been going down the rabbit hole for more than an hour and still can't make it work without disabling the sandbox. Is disabling the sandbox a no-go?

Reply:
Sorry for the question, but how did you run it in `x86_64`, under Docker? I'm having the same issue in my Fedora `x86_64` install.

Reply:
Sure, here are the steps (all run in `/spreed/`):
- `git fetch mishamosher`
- `git checkout mishamosher/chromium-recording`
- `docker rm --force --volumes talk-recording`
- `sed -i "s/options.add_argument('--no-sandbox')/#options.add_argument('--no-sandbox')/" recording/src/nextcloud/talk/recording/Participant.py`
- `recording/start-container.sh`
- _Ctrl+C_
- `docker exec talk-recording apt-get install --assume-yes software-properties-common`
- `docker exec talk-recording add-apt-repository ppa:phd/chromium-browser`
- `docker exec talk-recording bash -c 'printf "Package: *\nPin: release o=LP-PPA-phd-chromium-browser\nPin-Priority: 1001" > /etc/apt/preferences.d/phd-chromium-browser'`
- `docker exec talk-recording apt-get install --assume-yes chromium-browser chromium-chromedriver`
- Add the needed configuration in `/etc/nextcloud-talk-recording/server.conf` inside the container (listener, secrets, use _chrome_...)
- `recording/start-container.sh`
- In Talk, start a call and record it
- Fails. Ah, right, I forgot that my signaling server is using a self signed certificate, so I need to add `options.add_argument('--ignore-certificate-errors')` below `options = ChromeOptions()` (inside the container)
- Record the call again
- It works :shrug:

Reply:
Hmm...

Does the command `xvfb-run chromium-browser` spawn a non dying chromium
process?

Will try on a freshly installed Ubuntu 22.04 and Fedora 38 VMs on x86_64 to
discard anything related to my current setup.

Reply:
For the record, I have also tried with https://github.com/nextcloud/spreed/pull/9177 and it works too: 
- `git checkout mishamosher/chromium-recording`
- `sed -i "s/options.add_argument('--no-sandbox')/#options.add_argument('--no-sandbox')/" recording/src/nextcloud/talk/recording/Participant.py`
- `git cherry-pick mishamosher/docker-compose-recording`
- Apply the last changes described in https://github.com/nextcloud/spreed/pull/9177
- `cd recording/docker-compose/`
- `docker-compose up`
- In another terminal, add the needed configuration in `/etc/nextcloud-talk-recording/server.conf` inside the container: listener, secrets, use _chrome_...)
  - To allow the self signed certificate of my signaling server, add `options.add_argument('--ignore-certificate-errors')` below `options = ChromeOptions()` (inside the container)
- _Ctrl+C_
- `docker-compose up`
- Configure Talk to use the recording server (and the containers of the Nextcloud server and the HPB)
- In Talk, start a call and record it
- It works :shrug:

Reply:
Finished installing Ubuntu 22.04 and Fedora 38. Both x86_64, and up-to-date.

Tried many, many workarounds (as long as they don't disable the isolation already provider by Docker), but I am still unable to start Chromium inside Docker with sandboxing enabled.

```
  File "/usr/local/lib/python3.8/dist-packages/nextcloud/talk/recording/Participant.py", line 474, in __init__
    self.seleniumHelper.startChrome(width, height, env)
  File "/usr/local/lib/python3.8/dist-packages/nextcloud/talk/recording/Participant.py", line 250, in startChrome
    self.driver = ChromeDriver(
  File "/usr/local/lib/python3.8/dist-packages/selenium/webdriver/chrome/webdriver.py", line 84, in __init__
    super().__init__(
  File "/usr/local/lib/python3.8/dist-packages/selenium/webdriver/chromium/webdriver.py", line 104, in __init__
    super().__init__(
  File "/usr/local/lib/python3.8/dist-packages/selenium/webdriver/remote/webdriver.py", line 286, in __init__
    self.start_session(capabilities, browser_profile)
  File "/usr/local/lib/python3.8/dist-packages/selenium/webdriver/remote/webdriver.py", line 378, in start_session
    response = self.execute(Command.NEW_SESSION, parameters)
  File "/usr/local/lib/python3.8/dist-packages/selenium/webdriver/remote/webdriver.py", line 440, in execute
    self.error_handler.check_response(response)
  File "/usr/local/lib/python3.8/dist-packages/selenium/webdriver/remote/errorhandler.py", line 245, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.WebDriverException: Message: unknown error: Chrome failed to start: crashed.
  (unknown error: DevToolsActivePort file doesn't exist)
  (The process started from chrome location /usr/bin/chromium-browser is no longer running, so ChromeDriver is assuming that Chrome has crashed.)
Stacktrace:
#0 0x55ddc2bd6b23 <unknown>
#1 0x55ddc28ebb21 <unknown>
#2 0x55ddc2915e9a <unknown>
#3 0x55ddc2910461 <unknown>
#4 0x55ddc2957b73 <unknown>
#5 0x55ddc29570af <unknown>
#6 0x55ddc294da23 <unknown>
#7 0x55ddc291c662 <unknown>
#8 0x55ddc291dd7e <unknown>
#9 0x55ddc2c03c73 <unknown>
#10 0x55ddc2bb9dde <unknown>
#11 0x55ddc2bb97b0 <unknown>
#12 0x55ddc2bba5f5 <unknown>
#13 0x55ddc2bffabb <unknown>
#14 0x55ddc2bba9ae <unknown>
#15 0x55ddc2b9a9a4 <unknown>
#16 0x55ddc2bc5728 <unknown>
#17 0x55ddc2bc58d5 <unknown>
#18 0x55ddc2bd08ff <unknown>
#19 0x7f4303900609 start_thread
```

Reply:
I am not aware of having modifed anything in my Docker configuration to disable the isolation... but I will investigate. This is really weird :thinking: 

But it is unlikely that I will be able to do that this week, sorry :-(

Reply:
Just to be clear, I've used these ISOs:
- https://releases.ubuntu.com/22.04.2/ubuntu-22.04.2-desktop-amd64.iso
- https://download.fedoraproject.org/pub/fedora/linux/releases/38/Workstation/x86_64/iso/Fedora-Workstation-Live-x86_64-38-1.6.iso

...then went ahead for a minimal install in a fresh VM, updated the system (apt/dnf), and followed the [Install Docker Engine](https://docs.docker.com/engine/install/) instructions. Afterwards, practically the same exact steps that you've described.

Reply:
After testing with a Ubuntu 22.04 LiveUSB I have been able to reproduce the issue (not being able to run Chromium without `--no-sandbox` in a Docker container due to the `Failed to move to new namespace` error). I guess that at some point I modified some defaults for Docker in my development machine and forgot to revert the changes. Sorry for the noise.

However, I am still against launching Chromium with `--no-sandbox` from the recording server. The recording server can be run in a Docker container, but other container technologies might have different defaults that would allow a sandboxed Chromium to run, and in any case it can also be run directly on the host, without any container. The sandbox is a security feature of Chromium, and disabling it is needed to run it in a Docker container, but it is not a limitation imposed by the recording server itself (like other arguments like starting in kiosk mode). Therefore, the recording server should not disable the sandbox; working around that should be done in Docker instead.

It seems that [the default seccomp of the Docker container](https://docs.docker.com/engine/security/seccomp/) could be overriden to [allow some extra system calls needed by Chromium in sandboxed mode](https://blog.jessfraz.com/post/how-to-use-new-docker-seccomp-profiles/). The linked profile seems to be a bit _old_, so maybe some of the additional syscalls are no longer needed.

Another alternative seems to be [enabling user namespace cloning in the kernel](https://github.com/puppeteer/puppeteer/blob/main/docs/troubleshooting.md#setting-up-chrome-linux-sandbox). Unlike the previous option I guess that this one affects all the containers rather than just the specific container where Chromium will be run, so it might be less recommended (but unfortunately I am not familiar enough with these security features to weigh that).

Finally, the alternative used by the Docker images for Selenium is [`--no-sandbox`](https://github.com/SeleniumHQ/docker-selenium/blob/c6df1ab8dc6a5aca05c163c429a062ada1d79c51/NodeChrome/wrap_chrome_binary#L25). However, this is done at the container level, by adding a wrapper script that runs Chromium with that option by default, so the Selenium code does not need to worry about that.

The docker-compose for the recording server could include a note about having to use one of these alternatives to run Chromium. I guess it would be even fine to apply one of the alternatives (either the seccomp profile or the wrapper to add `--no-sandbox`, although unfortunately I am not sure which one would be better from a security point of view), given that by default it would not be possible to run Chromium in Docker otherwise.

But independently of that `--no-sandbox` should be removed from _Participant.py_; for me that (and fixing the typo I have just seen :-) ) is the only remaining thing needed to merge this pull request.

Reply:
Got it.